pre_commit/languages/docker.py

```diff
@@ -20,21 +21,30 @@
 
 def _is_in_docker() -> bool:
     try:
-        with open('/proc/1/cgroup', 'rb') as f:
-            return b'docker' in f.read()
+        with open('/proc/1/mountinfo', 'rb') as f:
+            return b'/containers/' in f.read()
     except FileNotFoundError:
         return False
 
 
 def _get_container_id() -> str:
-    # It's assumed that we already check /proc/1/cgroup in _is_in_docker. The
-    # cpuset cgroup controller existed since cgroups were introduced so this
-    # way of getting the container ID is pretty reliable.
-    with open('/proc/1/cgroup', 'rb') as f:
+    # It's assumed that we already check /proc/1/mountinfo in _is_in_docker.
+    with open('/proc/1/mountinfo', 'rb') as f:
+        hostname_mount = re.compile(
+            r"""
+            /containers
+            (/overlay-containers)?
+            /([a-z0-9]{64})
+            (/userdata)?
+            /hostname
+            """,
+            re.VERBOSE,
+        )
```

Comment:
binary file and text regex is a bit of a mismatch -- you should match the two (probably binary file and binary regex)

also `re.compile` in a local scope isn't all that beneficial -- perhaps move it to the module scope?

also I think this can be done without a regex and will be more performant (since the start/end seem pretty unique)?